t/op/magic.t

```diff
@@ -433,6 +433,77 @@ EOP
   }
 }
 
+# Check that assigning to $0 properly handles UTF-8-stored strings:
+{
+
+  # We want $e9_with_utf8_pv’s PV to be UTF-8-encoded because we need to
+  # test that Perl translates UTF-8-stored code points to plain octets when
+  # assigning to $0.
+  #
+  my $e9_with_utf8_pv = "\xe9";
+  utf8::upgrade($e9_with_utf8_pv);
+
+  # This will be the same logical code point as $e9_with_utf8_pv, but
+  # implemented in Perl internally as a raw byte rather than UTF-8.
+  #
+  my $e9_with_plain_pv = $e9_with_utf8_pv;
+  utf8::downgrade($e9_with_plain_pv);
+
+  $0 = $e9_with_utf8_pv;
+
+  # In case the assignment to $0 changed $e9_with_utf8_pv, ensure that
+  # it is still interally double-UTF-8-encoded:
+  #
+  utf8::upgrade($e9_with_utf8_pv);
+
+  is ($0, $e9_with_utf8_pv, 'compare $0 to UTF8-flagged');
+  is ($0, $e9_with_plain_pv, 'compare $0 to non-UTF8-flagged');
+
+  my $linux_cmdline_cr = sub {
+    open my $rfh, '<', "/proc/$$/cmdline" or skip "open: $!", 1;
+    return do { local $/; <$rfh> };
+  };
+
+  SKIP: {
+    skip "Test is for Linux, not $^O", 2 if $^O ne 'linux';
+
+    my $slurp = $linux_cmdline_cr->();
+    like( $slurp, qr<\A$e9_with_utf8_pv\0>, '/proc cmdline shows as expected (compare to UTF8-flagged)' );
+    like( $slurp, qr<\A$e9_with_plain_pv\0>, '/proc cmdline shows as expected (compare to non-UTF8-flagged)' );
+  }
+
+  my $name_unicode = "haha\x{100}hoho";
+  my $name_utf8 = $name_unicode;
```

Comment:
But then I find the jargon around this all bizarre, and more suitable to an obfuscated code contest.

Reply:
_Sigh …_

This is why I wanted to rename “UTF8”: too much confusion between what a Perl maintainer like yourself thinks of as a “UTF-8 string” versus what a Perl _user_ would think that term to mean (which is how I named this variable).

If we had different terms for Perl’s internal encoding versus logical, application-level encoding, there’d be no confusion in cases like this that “straddle” the two spheres.

Anyway. Would $name_utf8_bytes be clearer to you?

Reply:
I renamed as indicated.

Reply:
Instead of a being a Luddite around encode/decode upgrade/downgrade, I prefer to think of myself as someone who chose not to drink the kool-aide, and hence has remained relatively sane